src/xdp-app-info-linyaps.c

```diff
@@ -0,0 +1,269 @@
+                                       error);
+  g_assert (instance_id != NULL);
+
+  network = g_key_file_get_string (metadata,
+                                   LINYAPS_METADATA_GROUP_CONTEXT,
+                                   LINYAPS_METADATA_KEY_NETWORK,
+                                   error);
+  g_assert (network != NULL);
+
+  if (g_strcmp0 (network, "shared") == 0)
+    flags |= XDP_APP_INFO_FLAG_HAS_NETWORK;
+  flags |= XDP_APP_INFO_FLAG_REQUIRE_GAPPINFO;
+
+  app_info_linyaps = g_initable_new (XDP_TYPE_APP_INFO_LINYAPS,
+                                     NULL,
+                                     error,
+                                     "engine", LINYAPS_ENGINE_ID,
+                                     "flags", flags,
+                                     "id", app_id,
+                                     "instance", instance_id,
+                                     "sender", sender,
+                                     NULL);
+
+  return XDP_APP_INFO (g_steal_pointer (&app_info_linyaps));
+}
+
+/* this implementation refers to flatpak */
+static int
+open_linyaps_info (int      pid,
+                   GError **error)
+{
+  g_autofree char *root_path = NULL;
+  g_autofd int root_fd = -1;
+  g_autofd int info_fd = -1;
+
+  root_path = g_strdup_printf ("/proc/%u/root", pid);
+  root_fd = openat (AT_FDCWD, root_path, O_RDONLY | O_NONBLOCK | O_DIRECTORY | O_CLOEXEC | O_NOCTTY);
+  if (root_fd == -1)
+    {
+      if (errno == EACCES)
+        {
+          struct statfs buf;
+          if (statfs (root_path, &buf) == 0 &&
+              buf.f_type == 0x65735546) /* FUSE_SUPER_MAGIC */
+          {
+            g_set_error (error, XDP_APP_INFO_ERROR,
+                        XDP_APP_INFO_ERROR_WRONG_APP_KIND,
+                        "Not a linyaps (fuse rootfs)");
+            return -1;
+          }
+        }
+
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Unable to open %s",
+                  root_path);
+      return -1;
+  }
+
+  info_fd = openat (root_fd, ".linyaps", O_RDONLY | O_CLOEXEC | O_NOCTTY);
+  if (info_fd == -1)
+    {
+      if (errno == ENOENT)
+        {
+          g_set_error (error, XDP_APP_INFO_ERROR, XDP_APP_INFO_ERROR_WRONG_APP_KIND,
+                      "Not a linyaps (no .linyaps)");
+          return -1;
+        }
+
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                  "Unable to open application info file");
+      return -1;
+  }
+
+  return g_steal_fd (&info_fd);
+}
+
+static void
+xdp_app_info_linyaps_dispose(GObject *object)
+{
+  G_OBJECT_CLASS (xdp_app_info_linyaps_parent_class)->dispose (object);
+}
+
+static void
+xdp_app_info_linyaps_class_init(XdpAppInfoLinyapsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  object_class->dispose = xdp_app_info_linyaps_dispose;
+}
+
+static void
+xdp_app_info_linyaps_init(XdpAppInfoLinyaps *app_info_linyaps)
+{
+}
+
+XdpAppInfo *
+xdp_app_info_linyaps_new(const char *sender,
```

Comment:
```suggestion
xdp_app_info_linyaps_new (const char *sender,
```
Same thing in other places.